src/main/clojure/clara/rules/durability.clj

```diff
@@ -184,12 +184,15 @@
 ;;; indexed-session-memory-state for more details.
 (defrecord MemIdx [idx])
 
+;;; Same as MemIdx but specific to internal objects, such as Token or Element.
+(defrecord InternalMemIdx [idx])
```

Comment:
If we had both "internal" and "external" facts indexed under the same numbering schema and within the same lists, it would be more difficult/expensive to map them back during deserialization.  Currently the numbers used for external facts correspond directly to an indexed lookup in a sequential coll that is provided by the consumer via our API contract - as you've mentioned.  If we stored internal facts separately, we'd have to someone merge those facts with the external facts collection in a way that was sure to put all of them in their original positions.  I can't think of a fast way to do that and even if we had one, I think it would be error-prone and just more complex than needed.

The simplifying way to deal with it does seem to just be to use a different indexing for the different use-cases.